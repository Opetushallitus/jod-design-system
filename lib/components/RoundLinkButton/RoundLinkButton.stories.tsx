import type { Meta, StoryObj } from '@storybook/react';

import { RoundLinkButton } from './RoundLinkButton';

const meta = {
  title: 'Primitives/RoundLinkButton',
  component: RoundLinkButton,
  tags: ['autodocs'],
} satisfies Meta<typeof RoundLinkButton>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Base: Story = {
  args: {
    label: 'RoundLinkButton',
    component: ({ children, ...rootProps }) => (
      <a href="/#" {...rootProps}>
        {children}
      </a>
    ),
  },
};
