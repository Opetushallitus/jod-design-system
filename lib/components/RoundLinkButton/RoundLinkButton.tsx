interface LinkProps {
  className?: string;
  children: React.ReactNode;
}

type RoundLinkButtonLink = React.ComponentType<LinkProps>;

export interface RoundLinkButtonProps {
  /** Text shown on the link */
  label: string;
  /** Hide label */
  hideLabel?: boolean;
  /** Selected */
  selected?: boolean;
  /** CSS classname */
  className?: string;
  component: RoundLinkButtonLink;
  /** Icon shown on the link */
  icon: React.ReactNode;
}

/** Button component for single-step user actions. */
export const RoundLinkButton = ({
  label,
  hideLabel = false,
  selected = false,
  className,
  component: Component,
  icon,
  ...rest
}: RoundLinkButtonProps) => {
  return (
    <Component
      {...rest}
      aria-label={label}
      className={`${className ? className : ''} ds:group ds:inline-flex ds:min-w-[64px] ds:sm:min-w-[124px] ds:flex-col ds:items-center ds:justify-center ds:gap-2 ds:border ds:border-none ds:no-underline`.trim()}
    >
      <span
        aria-hidden
        className={`${selected ? 'ds:bg-accent ds:text-white' : 'ds:text-primary-gray ds:bg-bg-gray-2 ds:group-hover:text-accent'} ds:flex ds:size-[64px] ds:select-none ds:items-center ds:justify-center ds:self-center ds:rounded-full`}
      >
        {icon}
      </span>
      <span
        className={`${hideLabel ? 'ds:hidden' : ''} ds:text-button-md ${selected ? 'ds:text-accent' : 'ds:text-primary-gray'} ds:group-hover:text-accent ds:group-hover:underline`.trim()}
      >
        {label}
      </span>
    </Component>
  );
};
