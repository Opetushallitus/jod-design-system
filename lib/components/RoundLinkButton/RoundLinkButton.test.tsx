import { render, screen } from '@testing-library/react';
import { describe, expect, it, test } from 'vitest';

import { RoundLinkButton } from './RoundLinkButton';

const DummyLink = ({ children, ...rootProps }: { children: React.ReactNode }) => {
  return (
    <a href="/#" {...rootProps}>
      {children}
    </a>
  );
};

describe('Snapshot testing', () => {
  test('Default', () => {
    const { container } = render(
      <RoundLinkButton component={(props) => <DummyLink {...props} />} label="Default" icon="target" />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test('selected', () => {
    const { container } = render(
      <RoundLinkButton
        component={(props) => <DummyLink {...props} />}
        label="Selected"
        selected={true}
        icon="target"
      />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test('Disabled, selected', () => {
    const { container } = render(
      <RoundLinkButton
        component={(props) => <DummyLink {...props} />}
        label="Disabled, nonselected"
        selected={true}
        icon="target"
      />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test('Disabled, non-selected', () => {
    const { container } = render(
      <RoundLinkButton component={(props) => <DummyLink {...props} />} label="Disabled, nonselected" icon="target" />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test('Disabled, selected', () => {
    const { container } = render(
      <RoundLinkButton
        component={(props) => <DummyLink {...props} />}
        label="Disabled, nonselected"
        selected={true}
        icon="target"
      />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

it('has the correct label', () => {
  render(<RoundLinkButton component={(props) => <DummyLink {...props} />} label="label" icon="target" />);
  expect(screen.getByRole('link', { name: 'label' })).not.toBeNull();
});

it('renders with data-testid when provided', () => {
  render(
    <RoundLinkButton component={(props) => <DummyLink {...props} />} label="Has id" icon="target" dataTestId="rlbtn" />,
  );
  expect(screen.getByTestId('rlbtn')).toBeInTheDocument();
});
