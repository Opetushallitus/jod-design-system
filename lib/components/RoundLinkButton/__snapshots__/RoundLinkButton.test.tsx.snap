// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Snapshot testing > Default 1`] = `
<a
  aria-label="Default"
  class="ds:group ds:inline-flex ds:min-w-[64px] ds:sm:min-w-[124px] ds:flex-col ds:items-center ds:justify-center ds:gap-2 ds:border ds:border-none ds:no-underline"
  href="/#"
>
  <span
    aria-hidden="true"
    class="ds:text-black ds:bg-bg-gray-2 ds:group-hover:text-accent ds:flex ds:size-[64px] ds:select-none ds:items-center ds:justify-center ds:self-center ds:rounded-full"
  >
    target
  </span>
  <span
    class="ds:text-button-md ds:text-black ds:group-hover:text-accent ds:group-hover:underline"
  >
    Default
  </span>
</a>
`;

exports[`Snapshot testing > Disabled, non-selected 1`] = `
<a
  aria-label="Disabled, nonselected"
  class="ds:group ds:inline-flex ds:min-w-[64px] ds:sm:min-w-[124px] ds:flex-col ds:items-center ds:justify-center ds:gap-2 ds:border ds:border-none ds:no-underline"
  href="/#"
>
  <span
    aria-hidden="true"
    class="ds:text-black ds:bg-bg-gray-2 ds:group-hover:text-accent ds:flex ds:size-[64px] ds:select-none ds:items-center ds:justify-center ds:self-center ds:rounded-full"
  >
    target
  </span>
  <span
    class="ds:text-button-md ds:text-black ds:group-hover:text-accent ds:group-hover:underline"
  >
    Disabled, nonselected
  </span>
</a>
`;

exports[`Snapshot testing > Disabled, selected 1`] = `
<a
  aria-label="Disabled, nonselected"
  class="ds:group ds:inline-flex ds:min-w-[64px] ds:sm:min-w-[124px] ds:flex-col ds:items-center ds:justify-center ds:gap-2 ds:border ds:border-none ds:no-underline"
  href="/#"
>
  <span
    aria-hidden="true"
    class="ds:bg-accent ds:text-white ds:flex ds:size-[64px] ds:select-none ds:items-center ds:justify-center ds:self-center ds:rounded-full"
  >
    target
  </span>
  <span
    class="ds:text-button-md ds:text-accent ds:group-hover:text-accent ds:group-hover:underline"
  >
    Disabled, nonselected
  </span>
</a>
`;

exports[`Snapshot testing > Disabled, selected 2`] = `
<a
  aria-label="Disabled, nonselected"
  class="ds:group ds:inline-flex ds:min-w-[64px] ds:sm:min-w-[124px] ds:flex-col ds:items-center ds:justify-center ds:gap-2 ds:border ds:border-none ds:no-underline"
  href="/#"
>
  <span
    aria-hidden="true"
    class="ds:bg-accent ds:text-white ds:flex ds:size-[64px] ds:select-none ds:items-center ds:justify-center ds:self-center ds:rounded-full"
  >
    target
  </span>
  <span
    class="ds:text-button-md ds:text-accent ds:group-hover:text-accent ds:group-hover:underline"
  >
    Disabled, nonselected
  </span>
</a>
`;

exports[`Snapshot testing > selected 1`] = `
<a
  aria-label="Selected"
  class="ds:group ds:inline-flex ds:min-w-[64px] ds:sm:min-w-[124px] ds:flex-col ds:items-center ds:justify-center ds:gap-2 ds:border ds:border-none ds:no-underline"
  href="/#"
>
  <span
    aria-hidden="true"
    class="ds:bg-accent ds:text-white ds:flex ds:size-[64px] ds:select-none ds:items-center ds:justify-center ds:self-center ds:rounded-full"
  >
    target
  </span>
  <span
    class="ds:text-button-md ds:text-accent ds:group-hover:text-accent ds:group-hover:underline"
  >
    Selected
  </span>
</a>
`;
